### DISCUSSION TITLE
How to pass optional URL to Kingfisher 6's new KF shorthand?

### DISCUSSION THREAD

**Discussion by dhritzkiv** [onevcat/Kingfisher#1614] [Q&A] [answered]
Using the 'old' `UIImageView.kf.setImage` API, I was able to pass in an optional URL (a result of, for example, when a user never uploaded a photo to their profile), and rely on `placeholder` to set a placeholder/fallback image, even in the case of a missing image url:

old way:
```swift
imageView.kf.setImage(
    // `user.pictureUri` (`URL?`) is `nil` here
    with: user.pictureUri,
    placeholder: UIImage(named: "UserPlaceholderImage")
)
```

new way (expected):
```swift
KF.url(user.pictureUri)
            .placeholder(UIImage(named: "UserPlaceholderImage"))
            .set(to: imageView)
```

However, `KF#url` requires `URL` and does not work with `URL?`

---

I can work around this by re-writing the code as:

```swift
imageView.image = UIImage(named: "UserPlaceholderImage")

if let url = user.pictureUri {
    KF.url(url).set(to: imageView)
}
```

which, while still fairly clean, doesn't benefit from using KF's new chainable API for all my UIImageView needs

---

If this was a deliberate decision, that's okay, too! I was merely expecting the two APIs to be identical in their capabilities.

**Comment by onevcat** [comment, marked as answer]
Umm, it is a trade-off between strictness and flexibility. The reason is I was receiving a bunch of issues that complains the image is not loaded. And at last, I have to ask them to check whether the URL is `nil` or not again and again.

But on the other hand, I guess yes, you are right on it. In this particular case, flexibility is more important.

Converting a `non-optional` parameter to an optional one does not break the current API and I guess I can do it very soon. Stay tuned!

**Reply by onevcat** [reply]
Done in #1617 and I will find a good time to release it.
